gnucash/import-export/import-pending-matches.cpp

```diff
@@ -30,157 +30,69 @@
-{
-    g_return_if_fail (map);
-
-    g_hash_table_destroy (map);
-}
-
 static const GncGUID *
 gnc_import_PendingMatches_get_key (GNCImportMatchInfo *match_info)
 {
-    Split *split;
-    const GncGUID *match_guid;
-
     g_return_val_if_fail (match_info, NULL);
-
-    split = gnc_import_MatchInfo_get_split (match_info);
-    match_guid = qof_instance_get_guid (split);
-
-    return match_guid;
+    return qof_instance_get_guid (gnc_import_MatchInfo_get_split (match_info));
 }
 
-static GNCPendingMatches *
-gnc_import_PendingMatches_get_value (GNCImportPendingMatches *map,
+static GNCPendingMatches&
+gnc_import_PendingMatches_get_value (GNCImportPendingMatches& map,
                                      GNCImportMatchInfo *match_info)
 {
-    GNCPendingMatches *pending_matches;
-    const GncGUID *match_guid;
-
-    g_return_val_if_fail (map, NULL);
-    g_return_val_if_fail (match_info, NULL);
-
-    match_guid = gnc_import_PendingMatches_get_key (match_info);
-
-    pending_matches = g_hash_table_lookup (map, match_guid);
-
-    return pending_matches;
+    auto match_guid = gnc_import_PendingMatches_get_key (match_info);
+    return map[match_guid];
 }
 
 void
-gnc_import_PendingMatches_add_match (GNCImportPendingMatches *map,
+gnc_import_PendingMatches_add_match (GNCImportPendingMatches& map,
                                      GNCImportMatchInfo *match_info,
                                      gboolean selected_manually)
 {
-    GNCPendingMatches *pending_matches;
-    const GncGUID *match_guid;
-    GncGUID *key;
-
-    g_return_if_fail (map);
     g_return_if_fail (match_info);
 
-
-    pending_matches = gnc_import_PendingMatches_get_value (map, match_info);
-    match_guid = gnc_import_PendingMatches_get_key (match_info);
-
-    if (pending_matches == NULL)
-    {
-        pending_matches = g_new0 (GNCPendingMatches, 1);
-        key = g_new (GncGUID, 1);
-        *key = *match_guid;
-        g_hash_table_insert (map, key, pending_matches);
-    }
+    auto& pending_matches = gnc_import_PendingMatches_get_value (map, match_info);
 
     if (selected_manually)
-    {
-        pending_matches->num_manual_matches++;
-    }
+        pending_matches.num_manual_matches++;
     else
-    {
-        pending_matches->num_auto_matches++;
-    }
+        pending_matches.num_auto_matches++;
 }
 
 void
-gnc_import_PendingMatches_remove_match (GNCImportPendingMatches *map,
+gnc_import_PendingMatches_remove_match (GNCImportPendingMatches& map,
                                         GNCImportMatchInfo *match_info,
                                         gboolean selected_manually)
 {
-    GNCPendingMatches *pending_matches;
-
-    g_return_if_fail (map);
     g_return_if_fail (match_info);
 
-    pending_matches = gnc_import_PendingMatches_get_value (map, match_info);
-
-    g_return_if_fail (pending_matches);
+    auto& pending_matches = gnc_import_PendingMatches_get_value (map, match_info);
```

Comment:
So this is a problem: If `map` doesn't already have an entry consistent with `match_info` it will add one, and then decrement whichever of the matches counters to -1. It all works because you test for the counter being > 0 but there's a lot of misdirection to reason through.

Reply:
I think trying to decrement is a coding error. How about latest commit?

Reply:
I think adding an entry to the map from a function called `remove_match` is an error. So is mutating the container in a function named `get`: Getters should be `const`.

Reply:
ok rewritten to throw out_of_range if trying to decrement an inexistent pending_match